google/cloud/pubsub/internal/tracing_batch_sink.cc

```diff
@@ -0,0 +1,217 @@
+                          {/*sc::kMessagingOperation=*/
+                           "messaging.operation", "publish"},
+                          {sc::kThreadId, internal::CurrentThreadId()}},
+                         /*links*/ std::move(links), options);
+
+  // Add metadata to the message spans about the batch sink span.
+  auto context = batch_sink_parent->GetContext();
+  auto trace_id = internal::ToString(context.trace_id());
+  auto span_id = internal::ToString(context.span_id());
+  for (auto const& message_span : message_spans) {
+#if OPENTELEMETRY_ABI_VERSION_NO >= 2
+    message_span->AddEvent("gl-cpp.publish_start");
+    message_span->AddLink(context, {{}});
+#else
+    message_span->AddEvent("gl-cpp.publish_start",
+                           Attributes{{"gcp_pubsub.publish.trace_id", trace_id},
+                                      {"gcp_pubsub.publish.span_id", span_id}});
+#endif
+  }
+  return batch_sink_parent;
+}
+
+auto MakeChild(
+    opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span> const& parent,
+    int count, Links const& links) {
+  opentelemetry::trace::StartSpanOptions options;
+  options.parent = parent->GetContext();
+  options.kind = opentelemetry::trace::SpanKind::kClient;
+  return internal::MakeSpan("publish #" + std::to_string(count),
+                            /*attributes=*/{{}},
+                            /*links=*/links, options);
+}
+
+Spans MakeBatchSinkSpans(Spans const& message_spans, Options const& options) {
+  auto const max_otel_links = options.get<pubsub::MaxOtelLinkCountOption>();
+  Spans batch_sink_spans;
+  // If the batch size is less than the max size, add the links to a single
+  // span. If the batch size is greater than the max size, create a parent
+  // span with no links and each child spans will contain links.
+  if (message_spans.size() <= max_otel_links) {
+    batch_sink_spans.push_back(MakeParent(
+        MakeLinks(message_spans.begin(), message_spans.end()), message_spans));
+    return batch_sink_spans;
+  }
+  batch_sink_spans.push_back(MakeParent({{}}, message_spans));
+  auto batch_sink_parent = batch_sink_spans.front();
+
+  auto cut = [&message_spans, max_otel_links](auto i) {
+    auto const batch_size = static_cast<std::ptrdiff_t>(max_otel_links);
+    return std::next(
+        i, std::min(batch_size, std::distance(i, message_spans.end())));
+  };
+  int count = 0;
+  for (auto i = message_spans.begin(); i != message_spans.end(); i = cut(i)) {
+    // Generates child spans with links between [i, min(i + batch_size, end))
+    // such that each child span will have exactly batch_size elements or less.
+    batch_sink_spans.push_back(
+        MakeChild(batch_sink_parent, count++, MakeLinks(i, cut(i))));
+  }
+
+  return batch_sink_spans;
+}
+
+}  // namespace
+
+/**
+ * Records spans related to a batch messages across calls and
+ * callbacks in the `BatchingPublisherConnection`.
+ */
+class TracingBatchSink : public BatchSink {
```

Comment:
Nit:  It looks like this could also go into the anonymous namespace.

Reply:
Done